### ISSUE TITLE
Clarify expected remote execution's ActionResult nesting levels

### ISSUE THREAD

**Issue by illicitonion** [bazelbuild/bazel#4767]
As per https://groups.google.com/forum/#!topic/bazel-buildfarm/BopNFI0dRiA

`google.devtools.remoteexecution.v1test.ActionResult` is specified in https://github.com/googleapis/googleapis/blob/133a69477b59c7f56bc290fa9f81750acc218ba4/google/devtools/remoteexecution/v1test/remote_execution.proto but it is unclear how nested the root is expected to be. Here are two possible interpretations for an `output_directory` representing `dir` in the working directory which contains the output file `file` (flattening digests into `Directories`):

1: The output_directory's path exactly matches the `Action`'s `output_directories` literal, and the `tree` represents the *contents* of that directory:
```
output_directories: {
  path: "/dir"
  tree: {
    root:
      files: "file"
    }
  }
}
```

2: The output_directory's path is that of the parent of the `Action`'s `output_directories` literal, and the `tree` represents the directory which nests the contents inside it:
```
output_directories: {
  path: "/"
  tree: {
    children: {
      directories:"dir"
    }
  }
}
```

I propose that we specify 1; that the `path` value of each `output_directories` in the `ActionResult` must exactly match its corresponding `output_directories` literal in the `Action`.

cc @ulfjack @werkt @dgassaway

**Comment by ola-rozenfeld** [comment]
Agreed that we should clarify https://github.com/googleapis/googleapis/blob/master/google/devtools/remoteexecution/v1test/remote_execution.proto#L675, in particular replace "input root" with working directory, but all paths are explicitly relative to the working directory. They never begin with a slash. In the dir/file case, the output tree is:

output_directories: {
  path: "dir"
  tree_digest: "###" 
}

Tree {
    root:
      files: "file"
    }
}

No path can begin with a slash, in particular cannot consist solely of a slash.

**Comment by werkt** [comment]
This says that "/" is a valid path, and that trailing slashes are absolutely forbidden.  It doesn't indicate whether leading slashes are allowed:
https://github.com/googleapis/googleapis/blob/133a69477b59c7f56bc290fa9f81750acc218ba4/google/devtools/remoteexecution/v1test/remote_execution.proto#L326-L330

This refers to a message definition that does not exist, and gives no indication as to how it will be pathed, as though the population of the ActionResult output_directories will be index-linked to the output_directories from the Action:
https://github.com/googleapis/googleapis/blob/133a69477b59c7f56bc290fa9f81750acc218ba4/google/devtools/remoteexecution/v1test/remote_execution.proto#L589-L593

This indicates the "relative to the input root" behavior, but doesn't specifically omit leading slashes, and says nothing about trailing slashes like the others do:
https://github.com/googleapis/googleapis/blob/133a69477b59c7f56bc290fa9f81750acc218ba4/google/devtools/remoteexecution/v1test/remote_execution.proto#L675-L676

**Comment by ola-rozenfeld** [comment]
Oh, okay, I am surprised that we said "/" is allowed. I understand the
meaning was "capture the entire working directory tree, including the
inputs", but I suspect this should just be disallowed instead.
Alternatively, the path for this case should be empty instead of "/",
because "/" is misleading. And in any case, the output files and
directories for such an action output will be pathed relatively to the
working directory, as always. WDYT?

Agreed that we should copy-paste the restrictions everywhere, so that they
are consistent. The intention was for input and output trees to use the
same pathing scheme/restrictions.

**Comment by werkt** [comment]
I would prefer if we had a stringspec (or maybe another type) for these to conform to, and refer to in all three of these locations (plus wherever else they may exist) so that we DRY and don't end up with more transcription issues.

**Comment by ola-rozenfeld** [comment]
PTAL at https://docs.google.com/document/d/12c3oAPgedckLpue2yj0xGgJTEOyUm4mXWWBJ157J-8I/edit#heading=h.sg3la2izb6yt to address both this and https://github.com/bazelbuild/bazel/issues/4766
Thank you!

**Comment by illicitonion** [comment]
LGTM
